src/types/ianatimezone.jl

```diff
@@ -0,0 +1,157 @@
+const IANA_TABLE_SIZE = 1680
+
+perfect_hash(tz::VariableTimeZone) = perfect_hash(tz.name)
+function perfect_hash(str::AbstractString)
+    # This was generated via `gperf` and translated by hand.
+    # in case of collisions from new keys being added you must *not* regenerate it or
+    # change the assoc table, as that will change existing hashes and break any serialized
+    # data. Instead add special cases from the exact string via adding special cases. e.g.
+    #`name=="New/Timezone" && return 1681` (after adjusting `IANA_TABLE_SIZE`)
+    # The code below will (for all timezones in the 2021a release of tzdata) generate a
+    # value between 29 and 1680 inclusive. So any new keys that need to be added
+    # manually  to resolve collisions can freely use anything outside that range.
+
+    asso_values = (
+      1681, 1681, 1681, 1681, 1681, 1681, 1681, 1681, 1681,
+      1681, 1681, 1681, 1681, 1681, 1681, 1681, 1681, 1681, 1681,
+      1681, 1681, 1681, 1681, 1681, 1681, 1681, 1681, 1681, 1681,
+      1681, 1681, 1681, 1681, 1681, 1681, 1681, 1681, 1681, 1681,
+      1681, 1681, 1681,    4, 1681,    1,    3,    3,  241,   48,
+         1,   22,   10,    9,   61,   18,   74,    9,   30, 1681,
+      1681, 1681, 1681, 1681, 1681,    4,    2,   18,  167,   48,
+       547,  691,  532,  446,  574,  466,  211,  452,    8,  387,
+       110,  475,  601,  262,   83,  124,  580,  536,   60,    1,
+       719,    2, 1681, 1681, 1681,  291,    1,    1,   27,  424,
+        13,    5,   43,    3,  355,   12,  168,  148,   90,  179,
+         4,    1,  315,    3,    3,    3,    2,   37,    5,   65,
+        22,  320,  245,    1, 1681, 1681, 1681,
+    )
+
+    units = codeunits(str)
+    len = length(units)
+    hval = len
+
+    len >= 19 && (hval += asso_values[units[19]])
+    len >= 12 && (hval += asso_values[units[12]])
+    len >= 11 && (hval += asso_values[units[11]])
+    len >= 9 && (hval += asso_values[units[9] + 1])
+    len >= 8 && (hval += asso_values[units[8]])
+    len >= 6 && (hval += asso_values[units[6] + 1])
+    len >= 4 && (hval += asso_values[units[4]])
+    len >= 2 && (hval += asso_values[units[2] + 1])
+    len >= 1 && (hval += asso_values[units[1]])
+    len > 0 && (hval += asso_values[units[end]])  # add the last
+    return hval
+end
+
+
+const IANA_TIMEZONES = Vector{VariableTimeZone}(undef, IANA_TABLE_SIZE)
+
+# TODO: maybe fill this during build(), probably by generating a julia file.
```

Comment:
Or maybe it is fine, since it only queries what files exist, using `timezone_names` rather than reading them.
Maybe `timezone_names` (or `__init__`) should do the validation that TimeZones.jl has build properly and has the `TZDATA.COMPILED_DIR`.
Which would maybe clean up `TimeZone` a bit`.